### PR TITLE
highlevel: remove most of error handling from public API

### LINKED ISSUE #2430: Highlevel API tests
The highlevel API currently only does some basic tests for each of the supported types.

It would be nice, if we added more tests to `testkdb_highlevel.cpp` (especially for border cases).
For example adding a test that checks that we accept `0` and `65535`, but not `-1` and `65536` for the type `unsigned_short`.

### PR DESCRIPTION
**Pull request by kodebach**
As discussed in #2399, this removes most of the current error access functions from the public API.

It also adds border tests for integer types and therefore closes #2430.

@markus2330 please review my pull request

### CODE REVIEW

**Comment by kodebach** [comment]
jenkins build all please

**Comment by kodebach** [comment]
jenkins build all please

File: src/error/exporterrors.cpp [hunk 1]
```diff
@@ -524,6 +524,7 @@ static ostream & printPrivate (ostream & os, parse_t & p)
 	   << endl
 	   << "#include <elektra/types.h>" << endl
 	   << "#include <elektra/error.h>" << endl
+	   << "#include <kdbprivate.h>" << endl
```
Comment (markus2330) [inline src/error/exporterrors.cpp]: Why is this needed? It would be good if we can avoid it. If it is only about the enum, we better move the enum within here.

File: src/include/elektra/error.h [hunk 2]
```diff
@@ -20,34 +20,10 @@ extern "C" {
...
-	ELEKTRA_ERROR_SEVERITY_ERROR,
-	ELEKTRA_ERROR_SEVERITY_WARNING
-} ElektraErrorSeverity;
-
-typedef const char * ElektraKDBErrorGroup;
-typedef const char * ElektraKDBErrorModule;
-
 typedef void (*ElektraErrorHandler) (ElektraError * error);
 
 ElektraErrorCode elektraErrorCode (const ElektraError * error);
```
Comment (markus2330) [inline src/include/elektra/error.h]: Remove this also?

File: tests/kdb/testkdb_highlevel.cpp [hunk 3]
```diff
@@ -12,6 +12,7 @@
 
 #include <gtest/gtest-elektra.h>
 #include <kdbhelper.h>
+#include <kdbprivate.h>
```
Comment (markus2330) [inline tests/kdb/testkdb_highlevel.cpp]: We also should have ABI tests, i.e. tests that only use the public interface.

File: tests/kdb/testkdb_highlevel.cpp [hunk 4]
```diff
@@ -193,6 +200,141 @@ class Highlevel : public ::testing::Test
...
+		makeKey (KDB_TYPE_LONG_LONG, "longlong/above", "9223372036854775808"),
+		makeKey (KDB_TYPE_UNSIGNED_LONG_LONG, "unsignedlonglong/below", "-1"),
+		makeKey (KDB_TYPE_UNSIGNED_LONG_LONG, "unsignedlonglong/min", "0"),
+		makeKey (KDB_TYPE_UNSIGNED_LONG_LONG, "unsignedlonglong/max", "18446744073709551615"),
+		makeKey (KDB_TYPE_UNSIGNED_LONG_LONG, "unsignedlonglong/above", "18446744073709551616"),
+	});
+
+	createElektra ();
+
+	EXPECT_EQ (elektraGetOctet (elektra, "octet/min"), 0) << "octet/min wrong";
```
Comment (markus2330) [inline tests/kdb/testkdb_highlevel.cpp]: Thank you for adding the tests!

File: tests/kdb/testkdb_highlevel.cpp (hunk 3 above)
Comment (kodebach) [reply tests/kdb/testkdb_highlevel.cpp]: The tests themselves do only access the public interface. Some use the low-level API too to check the results, but `kdbprivate.h` is only used for the fatal error handler that outputs error messages. In particular we only use the `elektraError`* and `elektraKDBError*` functions that were part of the public interface before this PR.

File: src/error/exporterrors.cpp (hunk 1 above)
Comment (kodebach) [reply src/error/exporterrors.cpp]: It's needed for `ElektraErrorSeverity`, which is used in the generated function `elektraErrorLowLevel`. The function is used to create the `ElektraError` struct for low-level errors.

I don't think it is a problem to include `kdbprivate.h` here. The generated file is `elektra/errorsprivate.h` and is not intended for external use. It is not even included in any CMake install targets. It is not referenced in any header files, so it basically doesn't exist outside of compiling Elektra.

**Comment by kodebach** [comment]
> Is the high-level API ready to be released after this PR?

I was thinking, now that we decided to use the `specload` plugin for code generation, the `defaults` parameter of `elektraOpen` causes more problems than it solves. We might want to remove it. But I wouldn't actually remove the parameter. I would just not use it and rename it to `config`. That way we are prepared, if we need to pass more configuration parameters to `elektraOpen` in the future (e.g. enabling/disabling certain global plugins, or anything else that has to be done between `kdbOpen` and `kdbGet`).

**Comment by kodebach** [comment]
> That way we are prepared, if we need to pass more configuration parameters to `elektraOpen` in the future

Supporting the notification API through the highlevel API (including code generation) for example would require configuration between `kdbOpen` and `kdbGet`. In this case no reasonable defaults exist, unlike with cmd-line options, so configuration has to be user supplied.

We should also probably provide a `elektraGetKDB` function that returns the `KDB` handle used by the highlevel API, so that it can be combined with the low-level API more easily. Possibly we also expose the internal `KeySet`, but I am not sure about that. Although both of these can be added in a future release.

File: tests/kdb/testkdb_highlevel.cpp (hunk 3 above)
Comment (markus2330) [reply tests/kdb/testkdb_highlevel.cpp]: For ABI tests it would be needed that no methods of kdbprivate.h is used at all. Otherwise the ABI tests would also fail on changes within kdbprivate.h.

**Comment by markus2330** [comment]
> I was thinking, now that we decided to use the specload plugin for code generation, the defaults parameter of elektraOpen causes more problems than it solves.

The default parameter is still needed because also kdbOpen() or some unrelated part within kdbGet() can fail. The goal would be that the application is always be able to start, independent of KDB.

> Supporting the notification API through the highlevel API (including code generation) for example would require configuration between kdbOpen and kdbGet.

You mean for elektraIoSetBinding? I don't think that "configuration" would be an elegant solution, it seems to be more straight-forward to call functions because you need to pass function pointers.

> [get internal KDB and KeySet] Although both of these can be added in a future release.

Yes, let us only add it later when it is clear that it will be needed. (e.g. if we decide not to wrap functions like elektraIoSetBinding)

**Comment by kodebach** [comment]
> I don't think that "configuration" would be an elegant solution, it seems to be more straight-forward to call functions because you need to pass function pointers.

Yes, passing any sort of non-string data is very ugly via a KeySet, but I don't see a better solution other than using a custom struct which has to be allocated again. Any configuration before `kdbOpen` or before `kdbGet` has to be done by `elektraOpen`, so the user has to give that information somehow. We also can't delay the `kdbGet` call until the first `elektraGet*` because then `elektraGet*` could fail.

**Comment by kodebach** [comment]
> The default parameter is still needed because also kdbOpen() or some unrelated part within kdbGet() can fail. The goal would be that the application is always be able to start, independent of KDB.

Currently `elektraOpen` always returns `NULL` if `kdbOpen` or `kdbGet` fails.

**Comment by markus2330** [comment]
Yes, for severe problems elektraOpen is allowed to fail. But for a later `elektraGet` also `kdbGet` (or even `kdbOpen` in case of new mountpoints) might fail. But then we should continue to run, as we do not want to fail during run-time.

**Comment by kodebach** [comment]
jenkins build all

**Comment by kodebach** [comment]
jenkins build all please

**Comment by kodebach** [comment]
jenkins build all please

**Comment by kodebach** [comment]
Contrary to what github thinks, all build jobs where successful. One of the Cirrus CI jobs is listed twice, clicking on Details and viewing on Cirrus CI reveals that the job finished.

**Comment by markus2330** [comment]
Thank you for the great work!
